destination_test.go

```diff
@@ -15,92 +30,150 @@ import (
 	"github.com/pinecone-io/go-pinecone/pinecone"
 )
 
-func destConfigFromEnv() DestinationConfig {
+func destConfigFromEnv(t *testing.T) DestinationConfig {
 	return DestinationConfig{
-		PineconeAPIKey:  os.Getenv("API_KEY"),
-		PineconeHostURL: os.Getenv("HOST_URL"),
+		APIKey: requiredEnv(t, "API_KEY"),
+		Host:   requiredEnv(t, "HOST_URL"),
 	}
 }
 
-func TestDestination_Integration_Insert(t *testing.T) {
-	is := is.New(t)
+func requiredEnv(t *testing.T, key string) string {
+	val := os.Getenv(key)
+	if val == "" {
+		t.Fatalf("env var %v unset", key)
+	}
+
+	return val
+}
+
+func TestDestination_NamespaceSet(t *testing.T) {
 	ctx := context.Background()
+	destCfg := destConfigFromEnv(t)
+	// we use the default namespace on all other tests, so it's correct to set it here
+	destCfg.Namespace = "test-namespace"
 
-	dest := NewDestination()
+	is := is.New(t)
+	dest := newDestination()
 
-	destCfg := destConfigFromEnv()
+	err := dest.Configure(ctx, destCfg.toMap())
+	is.NoErr(err)
+
+	err = dest.Open(ctx)
+	is.NoErr(err)
+	defer teardown(ctx, is, dest)
+
+	is.Equal(dest.writer.index.Namespace, "test-namespace")
+}
+
+func TestDestination_Integration_WriteDelete(t *testing.T) {
+	ctx := context.Background()
+	destCfg := destConfigFromEnv(t)
+	is := is.New(t)
+	dest := newDestination()
 
 	err := dest.Configure(ctx, destCfg.toMap())
 	is.NoErr(err)
 
 	err = dest.Open(ctx)
 	is.NoErr(err)
-	defer teardown(is, ctx, dest)
+	defer teardown(ctx, is, dest)
 
 	id := uuid.NewString()
-	payload, err := json.Marshal(recordPayload{
+	position := sdk.Position(fmt.Sprintf("pos-%v", id))
+	metadata := map[string]string{
+		"pinecone.prop1": "val1",
+		"pinecone.prop2": "val2",
+	}
+
+	vecsToBeWritten := recordPayload{
+		ID:     id,
 		Values: []float32{1, 2},
 		SparseValues: sparseValues{
-			Indices: []uint32{1, 2},
-			Values:  []float32{1, 2},
+			Indices: []uint32{3, 5},
+			Values:  []float32{0.5, 0.3},
 		},
-	})
-	is.NoErr(err)
-
-	metadata := map[string]string{
-		"prop1": "val1",
-		"prop2": "val2",
 	}
 
-	rec := sdk.Util.Source.NewRecordCreate(
-		sdk.Position(fmt.Sprintf("pos-%v", id)),
-		metadata,
-		sdk.RawData(id),
-		sdk.RawData(payload),
-	)
+	payload, err := json.Marshal(vecsToBeWritten)
+	is.NoErr(err)
 
+	rec := sdk.Util.Source.NewRecordCreate(position, metadata, sdk.RawData(id), sdk.RawData(payload))
 	_, err = dest.Write(ctx, []sdk.Record{rec})
 	is.NoErr(err)
 
-	index := createIndex(is)
-	assertWrittenRecord(is, ctx, index, id, rec)
-}
-
-func createIndex(is *is.I) *pinecone.IndexConnection {
-	destCfg := destConfigFromEnv()
+	assertWrittenRecordIndex(ctx, t, is, dest.writer.index, id, vecsToBeWritten)
```

Comment:
Regarding `dest.writer.index`: I'd suggest creating a new client, just for the tests. Tests should work something like an independent auditor. If they rely on the same "tools" as the entity being audited, they might get wrong results. In our case, if the writer was set up incorrectly (e.g. writing to a different host, namespace, etc.), this test will pass, even though it's a complete failure.

Reply:
Fair, good point :+1: